### PR TITLE
feat(ec2): support using ssm parameter to resolve AMI ID at instance launch time

### PR DESCRIPTION
**Pull request by pahud**
Launch Template and EC2 instance support using SSM parameter to resolve the AMI ID at instance launch time(`resolve:ssm:parameter`) rather than the CFN deploy time(`CfnDynamicReference`). This PR introduces a new support for that.

- [Using SSM Parameter with Autoscaling and Launch Template](https://docs.aws.amazon.com/autoscaling/ec2/userguide/using-systems-manager-parameters.html)
- [Launch an instance using a Systems Manager parameter](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/finding-an-ami.html#using-systems-manager-parameter-to-find-AMI)

Remove `latestAmazonLinux2022()` from the integ test as it does not return any valid al2022 images anymore as described in https://github.com/aws/aws-cdk/issues/26274

Closes https://github.com/aws/aws-cdk/issues/24551

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts [hunk 1]
```diff
@@ -176,7 +191,7 @@ export abstract class MachineImage {
 }
 
 /**
- * Select the image based on a given SSM parameter
+ * Select the image based on a given SSM parameter at deploy time.
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:194]: Nice! But maybe we can make this even more clear:
```suggestion
 * Select the image based on a given SSM parameter at deployment time of the CloudFormation Stack.
```

File: packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts [hunk 2]
```diff
@@ -250,10 +299,17 @@ export interface SsmParameterImageOptions {
...
+   * The version of the SSM parameter.
+   *
+   * @default no version specified.
+   */
+  readonly parameterVersion?: string;
 }
 
 /**
- * Select the image based on a given SSM parameter
+ * Select the image based on a given SSM parameter at deploy time.
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:312]:
```suggestion
 * Select the image based on a given SSM parameter at deployment time of the CloudFormation Stack.
```

File: packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts [hunk 3]
```diff
@@ -156,6 +156,21 @@ export abstract class MachineImage {
...
+   *
+   * The AMI ID will be resolved at instance launch time.
+   *
+   * @param parameterName The name of SSM parameter containing the AMI ID
+   * @param options The parameter image options
+   *
+   * @see https://docs.aws.amazon.com/autoscaling/ec2/userguide/using-systems-manager-parameters.html
+   *
+   */
+  public static fromSsmParameterResolve(parameterName: string, options?: SsmParameterImageOptions): IMachineImage {
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:170]: Maybe something like this to make it more explicit?
```suggestion
  public static resolveSsmParameterAtLaunch(parameterName: string, options?: SsmParameterImageOptions): IMachineImage {
```

**Review by mrgrain — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts [hunk 4]
```diff
@@ -212,6 +227,40 @@ export class GenericSSMParameterImage implements IMachineImage {
...
+ * Select the image based on a given SSM parameter at instance launch time.
+ *
+ * This Machine Image comes with an imageId as `resolve:ssm:parameter-name` or `resolve:ssm:parameter-name:version` format
+ * as described in the document:
+ *
+ * @see https://docs.aws.amazon.com/autoscaling/ec2/userguide/using-systems-manager-parameters.html
+ *
+ * The AMI ID would be selected at instance launch time.
+ */
+export class SsmParameterResolveImage implements IMachineImage {
```
Comment (mrgrain) [inline packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:240]: Same as above

File: packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts (hunk 3 above)
Comment (pahud) [reply packages/aws-cdk-lib/aws-ec2/lib/machine-image/machine-image.ts:170]: Yes I considered this option as well. Let me update it.

**Review by mrgrain — APPROVED** [review]

**Comment by racheliurui** [comment]
Any code examples to use this feature? Thanks

**Comment by robertd** [comment]
@racheliurui 

```
const launchTemplate = new ec2.LaunchTemplate(stack, 'LaunchTemplate', {
  machineImage: ec2.MachineImage.resolveSsmParameterAtLaunch('parameterName');
});
```

**Comment by robertd** [comment]
@racheliurui

Make sure your SSM Param is of `aws:ec2:image` DataType (see: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ssm-parameter.html) and not of `text`.

**Pull request merged**